### PR TITLE
Feature/confluence

### PR DESCRIPTION
**Pull request by ghost**
This is a monster first PR for basic Confluence publishing.  It will be hidden behind a switch to only selectively support some internal testing as we want to get a better sense of what is remaining before we can call it done.  

This is almost completely isolated but I will call out a couple of changes that were needed in shared code for review.

### CODE REVIEW

File: src/command/publish/cmd.ts [hunk 1]
```diff
@@ -172,6 +172,13 @@ async function publishAction(
     publishTarget?: PublishRecord,
     account?: AccountToken,
   ) => {
+    // enforce requiresRender
+    if (publishProvider.requiresRender && publishOptions.render === false) {
```
Comment (allenmanning) [inline src/command/publish/cmd.ts:176]: @dragonstyle Can you please review?

File: src/command/publish/deployment.ts [hunk 2]
```diff
@@ -214,8 +214,7 @@ export async function chooseDeployment(
       try {
         const targetOrigin = new URL(deployment.target.url!).origin;
         if (
-          originCounts.get(targetOrigin) === 1 &&
-          deployment.provider.listOriginOnly
+          originCounts.get(targetOrigin) === 1 && (deployment.provider?.listOriginOnly ?? false)
```
Comment (allenmanning) [inline src/command/publish/deployment.ts:217]: @dragonstyle Can you please review?

File: src/publish/common/publish.ts [hunk 3]
```diff
@@ -33,6 +33,7 @@ import { formatResourcePath } from "../../core/resources.ts";
 import { encodeAttributeValue } from "../../core/html.ts";
 import { haveArrowKeys } from "../../core/platform.ts";
 import { capitalizeWord } from "../../core/text.ts";
+import { RenderFlags } from "../../command/render/types.ts";
```
Comment (allenmanning) [inline src/publish/common/publish.ts:36]: @dragonstyle Can you please review this file?

File: src/publish/README.md [hunk 4]
```diff
@@ -1,108 +0,0 @@
-# Quarto Publish
```
Comment (allenmanning) [inline src/publish/README.md:1]: @allenmanning I need to restore this.

**Review by ghost — COMMENTED** [review]
@dragonstyle Can I please request that three files with smaller changes be reviewed before I do the monster merge?  Any feedback on the other areas would be appreciated as well but it may be best for us to schedule something outside of github for that.

File: src/command/publish/cmd.ts (hunk 1 above)
Comment (allenmanning) [reply src/command/publish/cmd.ts:176]: This was from JJ's first session.

File: src/command/publish/deployment.ts (hunk 2 above)
Comment (allenmanning) [reply src/command/publish/deployment.ts:217]: This was from JJ's first session.

File: src/publish/common/publish.ts [hunk 5]
```diff
@@ -237,14 +238,14 @@ export async function handlePublish<
...
-  render: (siteUrl?: string) => Promise<PublishFiles>,
+  render: (flags?: RenderFlags) => Promise<PublishFiles>,
   providerName: string,
   type: "document" | "site",
   title: string,
   siteUrl?: string,
 ) {
   // render
-  let publishFiles = await render(siteUrl);
+  let publishFiles = await render({ siteUrl });
```
Comment (allenmanning) [inline src/publish/common/publish.ts:248]: I'm a little worried about this one.  I'm checking it out...

Comment (allenmanning) [reply src/publish/common/publish.ts:248]: It seems good.  The GH pages reference seems to use the flags as well.  

https://github.com/quarto-dev/quarto-cli/blob/0da4a0f5acc07291d9cbccb6ce78d78f9ef6dbf5/src/publish/gh-pages/gh-pages.ts#L105

File: src/core/url.ts [hunk 6]
```diff
@@ -7,6 +7,10 @@
 
 import { ensureTrailingSlash } from "./path.ts";
 
+export function isHttpUrl(url: string) {
+  return /^https?:/i.test(url);
+}
```
Comment (dragonstyle) [inline src/core/url.ts:12]: Probably not necessary but could consider `URL(url).protocol`...

File: src/publish/common/publish.ts (hunk 3 above)
Comment (dragonstyle) [reply src/publish/common/publish.ts:36]: done lgtm

File: src/resources/extensions/quarto/confluence/publish.lua [hunk 7]
```diff
@@ -0,0 +1,283 @@
+-- Confluence Storage Format for Pandoc
+-- https://confluence.atlassian.com/doc/confluence-storage-format-790796544.html
+-- https://pandoc.org/MANUAL.html#custom-readers-and-writers
+
+local stringify = (require "pandoc.utils").stringify
+local confluence = require('overrides')
+
+-- From https://stackoverflow.com/questions/9168058/how-to-dump-a-table-to-console
+function dumpObject(o)
```
Comment (dragonstyle) [inline src/resources/extensions/quarto/confluence/publish.lua:9]: this might be omitted as they are probably coming from `overrides` at this point?

**Review by dragonstyle — APPROVED** [review]
Looks great, I did my best to scan all the code (a lot of the confluence specific stuff was new to me so I merely looked pretty high level). Looks great this is going to be an awesome addition to Quarto!

A couple small and optional suggestions.

File: src/publish/publish.ts [hunk 8]
```diff
@@ -223,7 +249,7 @@ export async function publishDocument(
         });
       } else {
         throw new Error(
-          `The specifed document (${document}) is not a valid quarto input file`,
+          `The specifed document (${document}) is not a valid quarto input file`
```
Comment (cscheid) [inline src/publish/publish.ts:252]: Typo here (`specifed`) though it seems it existed before.

File: src/resources/extensions/quarto/confluence/overrides.lua [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+              return '&quot;'
+            elseif in_attribute and x == "'" then
+              return '&#39;'
+            else
+              return x
+            end
+          end)
+end
+
+-- From https://stackoverflow.com/questions/9168058/how-to-dump-a-table-to-console
```
Comment (cscheid) [inline src/resources/extensions/quarto/confluence/overrides.lua:27]: we have `quarto.log.output`, FWIW.

File: src/resources/extensions/quarto/confluence/publish.lua (hunk 7 above)
Comment (cscheid) [reply src/resources/extensions/quarto/confluence/publish.lua:9]: That, and `quarto.log.output`?

**Comment by zachcp** [comment]
Uhhh.... @allenmanning  this is amazing. Looking forward to trying this out.

**Pull request merged**